Add draining functionality to Consumer

mthssdrbrg opened the pull request:
While `#unsubscribe!` will cancel any active subscriptions, it won't guarantee that, when using the `:async` delivery mode, any more messages will be "delivered" to the `handler` block after the method returns.
This can cause a race condition when one wants to wait for all "in-flight" messages (being tracked outside of Autobahn) to be processed before calling `#disconnect!`.
So the solution to this would be to wait for the `@delivery_pool` to be shutdown before returning from `#unsubscribe!`. The test for this is horrible, but I couldn't come up with anything nicer right now.

This change breaks backward compatibility as now `#unsubscribe!` can raise an error that the delivery thread pool couldn't be shutdown in time.

mthssdrbrg committed:
Stop deliver thread pool in Consumer#unsubscribe!

While `#unsubscribe!` will cancel any active subscriptions, it won't
guarantee that, when using the `:async` delivery mode, any more messages
will be delivered after the method returns.
This causes a pretty sneaky race condition when one wants to wait for
all "in-flight" messages (being tracked outside of Autobahn) has been
processed before calling `#disconnect!`, in that ones `handler` block
can be called after `#unsubscribe!` returns.

Fonsan commented on spec/integration/autobahn_intg_spec.rb:
```diff
@@ -319,6 +319,20 @@ def create_transport_system(name, options={})
+          end
+        end
+        unsubscribed.release
+        @consumer.unsubscribe!
```
Given that the block in subscribe is run in another thread `#ack` might be called before `#unsubscribe!`, I can't see how this test ensures that `#ack` is not called either before or after `#unsubscribe!` making it indeterministic?

Fonsan replied:
To clarify, the expectation `unsubscribed.has_queued_threads.should be_false` might be correct in both cases, but it seems random which order the methods are called in the test

mthssdrbrg replied:
The test is not to ensure whether `#ack` is called either before or after `#unsubscribe!`, as what I wanted to test was that there wouldn't be any more calls to the given `handler` after `#unsubscribe!` returned.

grddev commented on spec/integration/autobahn_intg_spec.rb:
```diff
@@ -319,6 +319,20 @@ def create_transport_system(name, options={})
+        unsubscribed.release
+        @consumer.unsubscribe!
+        unsubscribed.has_queued_threads.should be_false
+      end
```
I think the problem here is that you try to combine too many things into one test. Assuming you only wanted to assert what the test description says, you would essentially want to just increment a counter from subscribe, and then set the counter to zero right after the unsubscribe call. A test like that would obviously require that you concurrently enqueue a bunch of things, to give a time-window after the unsubscribe-call finished where new calls could take place. Such a test would obviously not be deterministic.

I think you could replace the test with some more specific tests:
 - waits for current delivery to finish. by making the subscribe body increment a counters on entry and exit, with a little sleep in the middle, you could assert that the two counters should be equal (and nonzero) afterwards.
 - interrupts a delivery if it takes too long. do a semaphore acquire like you do in this test, but don't release it.
 - verify that it terminates the subscription thread(s). could be implemented by collecting `Thread.current` from the delivery callback, and asserting that the resulting threads aren't alive after the unsubscribe call.

None of these tests really capture the full desired behavior, but for that you would need to implement unit tests where you could actually inject the thread pool and whatnot. That seems like a much bigger change, though.

mthssdrbrg replied:
Yeah, the test itself was really horrible and as you say, tried to do everything at once. I like your suggestions and think I managed to cover them all.

grddev commented on lib/autobahn/consumer.rb:
```diff
@@ -59,6 +59,13 @@ def unsubscribe!(timeout=30)
             subscription.cancel
           end
           @subscriptions = nil
+          @deliver.set(false)
```
I don't remember the Autobahn internals properly, so I don't remember if this deals with batches of messages or not. If the `@deliver` flag is checked between delivery of messages in the same batch, we definitely do not want to set it to false here. As is, the deliver method stops immediately when `@deliver` is cleared and doesn't care if there are further messages enqueued. If these are full messages this should be ok, as RabbitMQ is supposed to handle that. If they are part of batches, we might process half batches and "exit cleanly" with this change.

mthssdrbrg replied:
You're correct, with this change we might process a portion of a batch and then just call it a day, which wouldn't be a good thing at all. I've changed how this works in the latest version.

grddev commented on lib/autobahn/consumer.rb:
```diff
@@ -121,10 +120,17 @@ def setup!
+        else
+          break
+        end
+      end
```
Is there any real reason to use the public next method here? Couldn't we use the demultiplexer directly?
```ruby
while @deliver.get
  if (entry = @demultiplexer.take(timeout))
    handler.call(*entry)
  end
end
while (entry = @demultiplexer.take(0))
  handler.call(*entry)
end
```

mthssdrbrg replied:
Not from what I can tell. The only difference would that we wouldn't call `#setup!` all of the time :)

grddev replied:
It turns out that the `handler.call(*entry)` is about 60% slower than `handler.call(headers, message)` in my version of JRuby, so that's a reason for staying with that part of the verbosity at least.

mthssdrbrg replied:
I had a feeling that the splatting would be slower, but you beat me to the benchmark.

mthssdrbrg committed:
Add Consumer#drain!

Basically just moves the stopping of the delivery thread pool to it's
own method (which also makes the changes backward-compatible).
The reason for renaming `@deliver` to `@drain` is simply because it fits
in a lot better with what it's actually used for.

mthssdrbrg commented:
So this took a turn towards then end, and the "draining" functionality has been moved to a separate method (`#drain!`), as it seemed as a better fit, and it's now backwards compatible (modulo log messages).

grddev commented:
LGTM. The addition of the new `#drain!` method is much nicer.
